File: chapters/04-data-structures-and-complexity/java/SlidingWindowRateLimit.java
// Scenario: per-user API rate limit — max 100 requests per 60-second sliding window
// Demonstrates: deque of timestamps; prune stale on each check
// Trade-off: memory per active user vs fixed-window counter (cheaper, burstier)

import java.util.ArrayDeque;
import java.util.Deque;

final class SlidingWindowRateLimit {
    private final int maxRequests;
    private final long windowMillis;
    private final Deque<Long> timestamps = new ArrayDeque<>();

    SlidingWindowRateLimit(int maxRequests, long windowMillis) {
        this.maxRequests = maxRequests;
        this.windowMillis = windowMillis;
    }

    synchronized boolean allow(long nowMillis) {
        while (!timestamps.isEmpty() && nowMillis - timestamps.peekFirst() >= windowMillis) {
            timestamps.pollFirst();
        }
        if (timestamps.size() >= maxRequests) return false;
        timestamps.addLast(nowMillis);
        return true;
    }
}

File: docs/chapters/04-data-structures-and-complexity/go/sliding_window_rate_limit.go
// Scenario: per-user API rate limit — max 100 requests per 60-second sliding window
// Demonstrates: slice queue of timestamps with prune from front
// Trade-off: Redis ZSET sliding window for distributed; local deque for edge gateway

package main

type SlidingWindowRateLimit struct {
	max       int
	windowMs  int64
	timestamps []int64
}

func NewSlidingWindowRateLimit(max int, windowMs int64) *SlidingWindowRateLimit {
	return &SlidingWindowRateLimit{max: max, windowMs: windowMs}
}

func (r *SlidingWindowRateLimit) Allow(nowMs int64) bool {
	cutoff := nowMs - r.windowMs
	i := 0
	for i < len(r.timestamps) && r.timestamps[i] < cutoff {
		i++
	}
	r.timestamps = r.timestamps[i:]
	if len(r.timestamps) >= r.max {
		return false
	}
	r.timestamps = append(r.timestamps, nowMs)
	return true
}
